polar-core/src/vm.rs

```diff
@@ -523,31 +526,24 @@ impl PolarVirtualMachine {
         Ok(QueryEvent::None)
     }
 
-    /// Return true if there is nothing left to do.
-    pub fn is_halted(&self) -> bool {
-        self.goals.is_empty() && self.choices.is_empty()
-    }
-
     /// Push a goal onto the goal stack.
     pub fn push_goal(&mut self, goal: Goal) -> PolarResult<()> {
-        if self.goals.len() >= self.stack_limit {
-            return Err(error::RuntimeError::StackOverflow {
-                limit: self.stack_limit,
-            }
-            .into());
-        }
+        use VariableState::Unbound;
         match goal {
-            Goal::LookupExternal { call_id, .. } | Goal::NextExternal { call_id, .. } => {
-                assert!(matches!(
-                    self.variable_state(self.get_call_sym(call_id)),
-                    VariableState::Unbound
-                ), "The call_id result variables for LookupExternal and NextExternal goals must be unbound.");
+            _ if self.goals.len() >= self.stack_limit =>
```

Comment:
Stylistic change or functional change? If its the former, I find this more confusing.

Reply:
Stylistic. There's three non overlapping paths through the function so I thought it was more clear to have them all be in one block at the same indentation level

Reply:
Yeah, that's fair. It seems a little weird in that the conditions are unrelated to the thing that is actually being matched on, but I don't feel that strongly about it

Reply:
I changed it to an `if/else` chain because I agree using a `match` like that is kind of odd

Reply:
Oh I love it. Good solution